d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeLoadBalancerStrategy.java

```diff
@@ -0,0 +1,109 @@
+*/
+
+package com.linkedin.d2.balancer.strategies.relative;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import com.linkedin.d2.balancer.strategies.ClientSelector;
+import com.linkedin.d2.balancer.strategies.LoadBalancerStrategy;
+import com.linkedin.d2.balancer.strategies.StateUpdater;
+import com.linkedin.d2.balancer.util.hashing.HashFunction;
+import com.linkedin.d2.balancer.util.hashing.Ring;
+import com.linkedin.r2.message.Request;
+import com.linkedin.r2.message.RequestContext;
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This strategy balances traffic to hosts within a service by dynamically adjusting a server's
+ * health score based on call statistics compared relatively to the performance of the entire cluster.
+ *
+ * Health score is rated on a scale from 0.0 - 1.0, with 0.0 meaning most unhealthy (all traffic
+ * routed away) and 1.0 meaning most healthy (no traffic routed away). Note that this behavior is
+ * inverse of dropRate in the degrader strategy.
+ *
+ * @see com.linkedin.d2.D2RelativeStrategyProperties
+ */
+public class RelativeLoadBalancerStrategy implements LoadBalancerStrategy
+{
+  private static final Logger LOG = LoggerFactory.getLogger(RelativeLoadBalancerStrategy.class);
+  public static final String RELATIVE_LOAD_BALANCER_STRATEGY_NAME = "relative";
+
+  private final StateUpdater _stateUpdater;
+  private final ClientSelector _clientSelector;
+
+  public RelativeLoadBalancerStrategy(StateUpdater stateUpdater,
+                                      ClientSelector clientSelector)
+  {
+    _stateUpdater = stateUpdater;
+    _clientSelector = clientSelector;
+  }
+
+  @Override
+  public String getName()
+  {
+    return RELATIVE_LOAD_BALANCER_STRATEGY_NAME;
+  }
+
+  @Nullable
+  @Override
+  public TrackerClient getTrackerClient(Request request,
+                                        RequestContext requestContext,
+                                        long clusterGenerationId,
+                                        int partitionId,
+                                        Map<URI, TrackerClient> trackerClients)
+  {
+    if (trackerClients == null || trackerClients.size() == 0)
+    {
+      LOG.warn("getTrackerClient called with null/empty trackerClients, so returning null");
+      return null;
+    }
+
+    _stateUpdater.updateState(new HashSet<>(trackerClients.values()), partitionId, clusterGenerationId);
+    Ring<URI> ring = getRing(clusterGenerationId, partitionId, trackerClients);
+    return _clientSelector.getTrackerClient(request, requestContext, ring, trackerClients);
+  }
+
+  @Nonnull
+  @Override
+  public Ring<URI> getRing(long clusterGenerationId, int partitionId, Map<URI, TrackerClient> trackerClients)
+  {
+    Ring<URI> ring = _stateUpdater.getRing(partitionId);
+    if (ring == null)
```

Comment:
That's a good point. The state will keep updating itself in every interval, because it is scheduled by the executor in the constructor of the state updater. So as long as it is initialized, it will still be updated. But one case I didn't take care of is the cluster generation id change, I will just always invoke updateState then.